modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/qa/matchers/source/MappingTransforms.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.datastreams.logsdb.qa.matchers.source;
+
+import java.util.HashMap;
+import java.util.Map;
+
+class MappingTransforms {
+    /**
+     * Normalize mapping to have the same structure as normalized source and enable field mapping lookup.
+     * Similar to {@link SourceTransforms#normalize(Map)} but needs to get rid of intermediate nodes
+     * and collect results into a different data structure.
+     *
+     * @param map
+     * @return
+     */
+    public static Map<String, Map<String, Object>> normalizeMapping(Map<String, Object> map) {
+        var flattened = new HashMap<String, Map<String, Object>>();
```

Comment:
I'm somewhat puzzled here, why not just use HashMap<String, Object> to track just the leaf fields?

Reply:
I see, you do that for source checking while you also want to compare the mappings. Need to think a bit about it..

Reply:
This is a map from normalized field name (`a.b.c`) to a map of mapping parameters (like `type`) since there can be multiple.